rules/jsll/src/rules/jsll-awa-init/jsll-awa-init.ts

```diff
@@ -38,73 +40,91 @@ const rule: IRuleBuilder = {
-                        }
                     }
                 };
             }
         });
 
         const validateScript = async (scriptParse: IScriptParse) => {
-            if (!hasJSLLScript || !isHead) {
+            const sourceCode = scriptParse.sourceCode;
+
+            if (!isPotentialInitScript(sourceCode)) {
+                immediateAfterJSLLscript = false;
+
+                return;
+            }
+
+            if (!traverseStarts) {
+                // When jsll init script is included in an external script
+                // `parse::javascript` of the init script is emitted before `element::script` of the JSLL link.
+                // So wait to run `validateScript` until traversal starts, otherwise the `immediateAfterJSLLscript` flag is not right.
+                tempInit = scriptParse;
+
                 return;
             }
 
-            const results = linter.verify(scriptParse.sourceCode, { rules: { 'jsll-awa-init': 'error' } });
+            if (isHead && !immediateAfterJSLLscript) {
+                await context.report(scriptParse.resource, null, `The JSLL init script should be immediately following the JSLL script.`);
+            }
+
+            if (!isHead) {
+                await context.report(scriptParse.resource, null, `The JSLL init script should be in <head>.`);
+            }
 
-            hasJSLLScript = false;
-            // Only validates the script included immediately after the JSLL link.
-            // So flip this flag to avoid duplicates.
+            immediateAfterJSLLscript = false;
+
+            const results = linter.verify(sourceCode, { rules: { 'jsll-awa-init': 'error' } });
 
             for (const result of results) {
                 await context.report(scriptParse.resource, null, result.message);
             }
         };
 
-        const enableInitValidate = (data: IElementFound) => {
+        const updateImmedateAfterJSLLScript = (data: IElementFound) => {
             const { element }: { element: IAsyncHTMLElement, resource: string } = data;
 
-            if (hasJSLLScript) {
+            if (immediateAfterJSLLscript) {
                 return;
             }
 
             // JSLL script has been included at this point.
             // Now JSLL init needs to be verified.
-            hasJSLLScript = isJsllDir(element);
+            immediateAfterJSLLscript = isJsllDir(element);
+
+            if (immediateAfterJSLLscript && tempInit) {
+                validateScript(tempInit);
+                tempInit = null;
+            }
+        };
+
+        const updateTraverseStart = () => {
+            traverseStarts = true;
         };
 
-        const enterBody = async (event: IElementFound) => {
-            const { resource }: { resource: string } = event;
+        const traverseUpHead = async (data: ITraverseUp) => {
```

Comment:
Can we subscribe to the `element::body` event instead of all the `traverse::up`?

This way you only receive this event once so things should run faster.

Reply:
I tried `element::body` in the beginning, it works in JSDOM, but in chrome the `element::body` gets fired before `parse::javascript` ends, and changing to `traverse::up` of head fixed it.

Reply:
> but in chrome the `element::body` gets fired before `parse::javascript` ends

This worries me. Events have to wait until all the handlers have been executed before continuing so this looks like an issue on the connector. Do you mind trying to solve that(several events being fired one finishes)? I can probably help you if needed later today.

Reply:
@molant I can definitely help but I'm afraid I won't get done with the JSLL before this afternoon. Currently making the fixes to enable as much validation as possible even if the location check fails. Let me know if you'd like me to look into the connector issue first.

Reply:
Finish all the JSLL stuff first and then we can see if we wait for a fix in sonarwhal for the events or not before merging.